Add OStreamSpanExporter

HudsonHumphries opened the pull request:
This PR implements the span half of the StreamExporter described in issue (https://github.com/open-telemetry/opentelemetry-cpp/issues/89). This exporter will not be used in production code, but rather in explorations or examples. This SpanExporter takes data in from the SpanProcessor in the form of recordables. The exporter then takes the recordables and gets all of the relevant data out of it and sends it to an ostream. The default configuration is to send the recordable data to stdout. The StreamMetricsExporter will be included in a future PR.

This PR includes tests for cout, cerr and clog as well as the shutdown function.

This PR also replaces the stub stdout_exporter in the simple tracing example (https://github.com/open-telemetry/opentelemetry-cpp/pull/86)

HudsonHumphries commented:
Tests build and pass when running locally and in docker container but fail in this CI jobs, looking into it

pyohannes commented on exporters/trace/stream/stream.h:
```diff
@@ -0,0 +1,180 @@
+{
+public:
+    explicit StreamSpanExporter(std::ostream &sout = std::cout,
+                                bool isShutdown = false) noexcept;
```
Can we add a doc string to the class and the constructor? I'm not clear about what the intention of the `isShutdown` argument is.

HudsonHumphries replied:
Yeah I'll add a doc string. I will remove isShutdown from the constructor since it should always be false when the exporter is created

pyohannes commented on exporters/trace/stream/stream.h:
```diff
@@ -0,0 +1,180 @@
+      case AttributeType::TYPE_STRING:
+        if (jsonTypes)
+          sout_ << '"';
+        // TODO: do we need to escape string value for JSON?
```
I'd say no. This exporter isn't expected to produce valid JSON output.

pyohannes commented on exporters/trace/stream/stream.h:
```diff
@@ -0,0 +1,180 @@
+  {
+    switch (value.index())
+    {
+      case AttributeType::TYPE_BOOL:
```
The only way I currently see to safely process `nostd::variant` values is via a visitor (like e. g. [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L36)).

I know @maxgolov has some valid concerns regarding that, and I'd appreciate him to weigh in.

With the current approach, any changes made to `common::AttributeValue` will break this code, at runtime. When using `nostd::visit`, changes to `common::AttributeValue` would cause a compiler error for this code and thus enable you to catch any incompatibilities.

pyohannes replied:
Also please rebase. `GetAttribute` on `SpanData` now returns a [SpanDataAttributeValue](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L22).

pyohannes replied:
You could also do something similar to what was done [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/176/files#diff-ca38b6b0446cadbf3f86cbf673da8783R24), but that doesn't warn you at compile time when the definition of the `variant` changed.

HudsonHumphries replied:
This way of printing Attributes is from one of @maxgolov branches. Will rebase to get updated Attributes

HudsonHumphries replied:
I rebased and am now using SpanDataAttributeValue, do you think this implementation is fine or should I go down a different path. I agree it will break if someone changes the variant, do you think that is going to happen often?

pyohannes replied:
I'd either go with [nostd::visitor](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L36) or [nostd::holds_alternative](https://github.com/open-telemetry/opentelemetry-cpp/pull/176/files#diff-ca38b6b0446cadbf3f86cbf673da8783R33).

HudsonHumphries replied:
I went with nostd::holds_alternative, let me know if you are okay with this implementation. I kept the print_array function in order to reduce code duplication.

pyohannes commented on exporters/trace/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+add_subdirectory(stream)
```
My biggest concern is that this will be confused with the term _streaming exporter_, which is something very different.

HudsonHumphries replied:
Does the subdirectory need to be the same as the folder name? Not sure what else to name it

reyang replied:
Discussed during the SIG meeting that we'll use `OStream`.

HudsonHumphries commented on examples/simple/main.cc:
```diff
@@ -3,20 +3,20 @@
 
 // Using an exporter that simply dumps span data to stdout.
-#include "stdout_exporter.h"
+#include "exporters/trace/stream/stream.h"
```
This is a much better way of doing it, appreciate it. Should be fixed now

HudsonHumphries commented:
Renamed to OStreamSpanExporter. Also added setting the TraceID / SpanID in tests so the tests do not break when IDs begin to be randomly generated. This is rebased with master as well

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+#include "opentelemetry/nostd/type_traits.h"
+#include <opentelemetry/version.h>
+
+#include <sstream>
```
nit: would you sort the name

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+#include "opentelemetry/sdk/trace/exporter.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/nostd/type_traits.h"
+#include <opentelemetry/version.h>
```
```suggestion
#include "opentelemetry/version.h"
```

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+public:
+
+/**
+ * Create a OStreamSpanExporter. This constructor takes in a reference to an ostream that the
```
```suggestion
 * Create an OStreamSpanExporter. This constructor takes in a reference to an ostream that the
```

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+    {7, "PERMISSION_DENIED"},
+    {8, "RESOURCE_EXHAUSTED"},
+    {9, "FAILED_PRECONDITION"},
+    {10,"ABORTED"},
```
nit: would you put a space here (if you like vertical alignment, put two spaces for single digit).

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+};
+}
+}
+OPENTELEMETRY_END_NAMESPACE
```
nit: add a blank line before EOF.

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+      sout_ << " ";
+    }
+
+    sout_ << "\n}\n";
```
Where is the corresponding `{`?

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,148 @@
+    {
+      sout_ << kv.first << ": ";
+      print_value(kv.second);
+      sout_ << " ";
```
need a `,`?

reyang commented on exporters/ostream/src/span_exporter.cc:
```diff
@@ -0,0 +1,69 @@
+              << "\n  description   : " << span->GetDescription() 
+              << "\n  status        : " << statusMap[int(span->GetStatus())]
+              << "\n  attributes    : ";
+              printAttributes(span->GetAttributes());
```
This seems hacky as  we rely on `printAttributes` to output the ending `}`.

reyang commented on exporters/ostream/test/ostream_span_test.cc:
```diff
@@ -0,0 +1,259 @@
+  "  duration      : 100\n"
+  "  description   : Test Description\n"
+  "  status        : UNIMPLEMENTED\n"
+  "  attributes    : attr1: [1,2,3] \n"
```
nit: try to avoid trailing space.

HudsonHumphries commented:
Thanks for the feedback @reyang let me know if I missed anything

reyang commented on exporters/ostream/include/opentelemetry/exporters/ostream/span_exporter.h:
```diff
@@ -0,0 +1,152 @@
+  void printAttributes(std::unordered_map<std::string, sdktrace::SpanDataAttributeValue> map)
+  {
+    int size = map.size();
+    int num = 1;
```
`num` is a confusing name.

HudsonHumphries replied:
Renamed to i. This branch is merged this master already

reyang commented:
#183

HudsonHumphries commented:
Design document can be found [here](https://github.com/open-o11y/docs/blob/master/exporter/cpp-ostream/ostream-exporter-design.md)
